Add Ostrom

kscholty opened the pull request:
In discussion https://github.com/evcc-io/evcc/discussions/15512 and closed issue https://github.com/evcc-io/evcc/issues/11148 people are asking for Ostrom support.
Since they have a public API and I had some spare time I implemented a tariff for EVCC.

The implementation supports both the SIMPLY_FAIR (fixed price for at least a month) and the SIMPLY_DYNAMIC (hourly rates) tariff, which would make it worth having a native implementation. 
The prices for both tariffs are queried on a regular basis.

Client credentials can be created under https://developer.ostrom-api.io/

Looking forward to your comments

kscholty committed:
Added ostrom tariff

kscholty commented:
I have not the slightest idea why the UI test fails. and how I can fix that. The same test shows nothing in my local repo.

kscholty committed:
New name for template

kscholty committed:
Deleted file for test purposes

kscholty committed:
Merge branch 'master' into tariff/ostrom

kscholty committed:
Revert "Deleted file for test purposes"

This reverts commit 0ab6018d217b950739c5f6ad0cbfb2def75b2e1d.

kscholty committed:
Linted template

kscholty commented:
@andig Warum draft? Nach meiner Ansicht ist das fertig so, oder hast du Änderungsvorschläge?

softcat commented:
Bezüglich des Preises für den SIMPLY_FAIR Tarif:

Ich nutze aktuell schlicht einen Query für die generische `tariff` Integration (ohne Authentifizierung)

```
https://api.ostrom.de/v1/tariffs/city-id?cityId=11111&usage=1000
```

Allerdings muss man sich dafür die `cityId` über die Homepage händisch raussuchen. Funktioniert dann aber gut. 

Gibt die API ggf. bzgl. der `cityId` etwas her? Ich kann es leider aktuell nicht testen. 

Ich wechsle in ein paar Wochen in den SIMPLE_DYNAMIC Tarif, dann ist deine Integration perfekt.

kscholty committed:
Automatic query of Simply Fai porices using
the v1 API.

kscholty commented:
Guter Hinweis. In der dokumentierten API gibt es nichts, um die CityId zu erfragen. Aber etwas suchen im Quelltext der Webseite ergab:
`https://api.ostrom.de/v1/addresses/cities?zip=${zip}`
Damit bekommt man die `CityId `zu einer Postleitzahl.
Ich habe den Code jetzt so angepasst, dass er das automatisch alles aus dem Contract ermittelt und abfragt.
Sollte sich der Simply Fair Preis also ändern, merkt evcc das spätestens an dem Tag, an dem es passiert. In die Zukunft abfragen geht leider nicht (bzw. weiß ich nicht genau, ab wann die API den neuen Preis liefert)

kscholty committed:
Merge branch 'master' into tariff/ostrom

kscholty committed:
Merge branch 'master' into tariff/ostrom

kscholty committed:
Merge branch 'master' into tariff/ostrom

kscholty committed:
Merge branch 'master' into tariff/ostrom
Removed DEBUG output

kscholty committed:
Merge branch 'tariff/ostrom' of https://github.com/kscholty/evcc into tariff/ostrom

userwithoutpassword commented:
Fehlt hier noch was?

kscholty commented:
> Fehlt hier noch was?

Aus meiner Sicht ist das so fertig. Hab's in der Sandbox mit beiden Varianten getestet und es funktioniert.

userwithoutpassword commented:
Wieso wird das jetzt kommentarlos geschlossen wenn sich jemand schon die Arbeit macht ne Integration zu basteln?

GrimmiMeloni commented:
Der Bot schließt nach 14 Tagen Inaktivität automatisch.

userwithoutpassword commented:
> > Fehlt hier noch was?
> 
> Aus meiner Sicht ist das so fertig. Hab's in der Sandbox mit beiden Varianten getestet und es funktioniert.

Kannst du nochmal drüber schauen woran es hakt?

kscholty committed:
Merge branch 'master' into tariff/ostrom

kscholty commented:
Da gibt's einen Test, der versucht einen Ostrom Tarif anzulegen und scheitert mit "unauthorized", was ich für korrekt halte. Keine Ahnung, wo der Test her kommt. Als ich das zuletzt eingecheckt habe, war alles grün. 
Kann mich frühestens nächste Woche darum kümmern,  bin im Urlaub.

GrimmiMeloni commented:
> Da gibt's einen Test, der versucht einen Ostrom Tarif anzulegen und scheitert mit "unauthorized", was ich für korrekt halte. Keine Ahnung, wo der Test her kommt. Als ich das zuletzt eingecheckt habe, war alles grün.


Das ist sehr wahrscheinlich ein Seiteneffekt von [#16528](https://github.com/evcc-io/evcc/pull/16528#issue-2568582769)

Wenn dieses Setup einen gültigen Account braucht, wird der im CI nicht durchlaufen. Die Lösung ist ein SkipTest attribut in der YAML hinzuzufügen. Beispiel für Tibber [hier](https://github.com/evcc-io/evcc/pull/16528/files#diff-6e81ce6f0766ec6defb1fbe66c363cf41f2de829e60707e94c05969d301e37b7).

kscholty committed:
Add "skip test"

kscholty committed:
Merge branch 'evcc-io:master' into tariff/ostrom

userwithoutpassword commented:
Kann es hinzugefügt werden? Würde es gern mal testen.

kscholty committed:
Merge branch 'master' into tariff/ostrom

kscholty committed:
Merge branch 'master' into tariff/ostrom

userwithoutpassword commented:
Kapier nicht so ganz wieso es hier nun 3mal "gemerged" wurde aber immer noch nicht unter "Comparing changes" auftaucht.

GrimmiMeloni commented:
> Kapier nicht so ganz wieso es hier nun 3mal "gemerged" wurde aber immer noch nicht unter "Comparing changes" auftaucht.

Was fehlt denn? Der Skiptest Change ist im Changeset zu sehen.

kscholty commented:
> Kapier nicht so ganz wieso es hier nun 3mal "gemerged" wurde aber immer noch nicht unter "Comparing changes" auftaucht.

Ich kann nur den aktuellen head in den branch mergen, was ich eben gemacht habe. Anders herum muss einer der maintainer machen. Es fehlt nichts mehr, aber ich glaube @andig möchte das nicht wirklich im Produkt haben, da er es für unnötig hält. 

Ich glaube allerdings, dass es sowohl im Marketing hilft, als auch dem Normal-User, der nicht versteht, wie er einen der anderen Tarife nutzen könnte.

userwithoutpassword commented:
Für die anderen Tarife gibt es doch ebenso templates.

GrimmiMeloni commented on util/element.go:
```diff
@@ -0,0 +1,44 @@
+			}
+		}
+	} else if len(objects) == 1 {
+		// vin empty and exactly one object
```
VIN? Copy and pasted maybe....

userwithoutpassword commented:
> Es fehlt nichts mehr, aber ich glaube @andig möchte das nicht wirklich im Produkt haben, da er es für unnötig hält.

Ich glaube die Entwickler stellen sich lieber stumm als eindeutig zu sagen was ist.
Einen Alternativen Weg wie man den Anbieter aktuell integriert sehe ich nicht...

kscholty commented:
Es hat ein Review gegeben mit ein paar Änderungswünschen. Habe es noch nicht geschafft, mir das anzuschauen. Mache ich in den nächsten Tagen. Hoffentlich geht es dann durch.

kscholty committed:
Merge branch 'evcc-io:master' into tariff/ostrom

kscholty commented on tariff/ostrom.go:
```diff
@@ -0,0 +1,271 @@
+	var once sync.Once
+
+	tick := time.NewTicker(time.Hour)
+	for ; true; <-tick.C {
```
That's hopw it is done in other tarrifs as well, so I did it the same way.

andig replied:
@GrimmiMeloni der for loop sorgt- im Gegensatz zu `range`- dafür, dass er sofort einmal durchlaufen wird. `select` sehe ich hier nicht?

GrimmiMeloni replied:
Sorry, hab mich vertan. Ich hatte an [dieses](https://gobyexample.com/tickers) Muster gedacht.

kscholty committed:
Implemented changes for review comments

GrimmiMeloni reviewed (approved):
Looking good to me.

@andig would you mind giving this a final glance and then merge?

kscholty committed:
Update tariff/ostrom.go

Co-authored-by: andig <cpuidle@gmail.com>

kscholty commented on util/element.go:
```diff
@@ -0,0 +1,44 @@
+	objects, err := list()
+
+	if err != nil {
+		return zero, fmt.Errorf("cannot get Objects: %w", err)
```
Das stimmt, denn die Funktion weiß ja nicht mehr, was sie da sucht. 
Alternativen sind: 
1) Man übergibt einen Parameter der entsprechend die Fehlermeldung anpasst
2) Eine vollständige Kopie der Funktion für generische Suchen und eine für Vehicles.

Erscheint mir beides schlechter als eine generische Fehlermeldung. Aber wenn es wichtig ist, kann ich gerne eine der Alternativen implementieren.

andig replied:
im EInzelfall hier wohl leichter neu zu impelmentieren ohne die komplexe Funktion.

kscholty replied:
Ok. Ich hatte nur gedacht, andere "nicht-Vehicle" könnten das so dann auch nutzen.

kscholty committed:
Apply suggestions from code review

Co-authored-by: andig <cpuidle@gmail.com>

kscholty committed:
Merge branch 'evcc-io:master' into tariff/ostrom

kscholty commented on tariff/ostrom.go:
```diff
@@ -0,0 +1,266 @@
+	}
+	done := make(chan error)
+
+	t.contractType = contract.Product
```
Geht nur mir größerem Umbau, da t selber die contracts abfragt (`t.GetContracts()`), die ich hier brauche, um den Wert zuzuweisen. Daher brauche ich vorher `t` und `t.transport`.

kscholty commented on tariff/ostrom.go:
```diff
@@ -0,0 +1,266 @@
+	return t, err
+}
+
+func addPrice(entry ostrom.ForecastInfo, rates api.Rates) api.Rates {
```
Ok, Wirkt tatsächlich schlanker so.

kscholty commented on tariff/ostrom.go:
```diff
@@ -0,0 +1,266 @@
+	}
+
+	uri := fmt.Sprintf("%s?%s", ostrom.URI_GET_CITYID, params.Encode())
+	if err := backoff.Retry(func() error {
```
Habs entfernt.

kscholty commented on tariff/ostrom.go:
```diff
@@ -0,0 +1,266 @@
+		t.log.ERROR.Println(err)
+		return 0, err
+	}
+	return city[0].Id, nil
```
Done

kscholty committed:
Implewmented more change requests.

andig commented on tariff/ostrom.go:
```diff
@@ -0,0 +1,266 @@
+	case ostrom.PRODUCT_FAIR, ostrom.PRODUCT_FAIR_CAP:
+		return api.TariffTypePriceStatic
+	default:
+		t.log.ERROR.Printf("Unknown tariff type %s\n", t.contractType)
```
Das Errorlogging. Der Tarif kann sich ja später nicht ändern.

kscholty replied:
Verstehe, Derzeit kann der default aber auch nicht erreicht werden. War mehr als Hinweis gedacht, falls die da mal was erweitern. Dafür halte ich eine Kopie des codes im init für overkill. Hab die Meldung daher ganz entfernt.

andig commented on tariff/ostrom.go:
```diff
@@ -0,0 +1,266 @@
+	client := request.NewHelper(t.log)
+	err := client.DoJSON(req, &res)
+
+	if err != nil {
```
Raus. Fehler nur 1x behandeln.

andig commented on tariff/ostrom.go:
```diff
@@ -0,0 +1,266 @@
+func (t *Ostrom) GetContracts() ([]ostrom.Contract, error) {
+	var res ostrom.Contracts
+
+	contractsURL := ostrom.URI_API + "/contracts"
+	err := t.GetJSON(contractsURL, &res)
```
```suggestion
	uri := ostrom.URI_API + "/contracts"
	err := t.GetJSON(uri, &res)
```

andig commented on tariff/ostrom.go:
```diff
@@ -0,0 +1,266 @@
+	var once sync.Once
+	var val ostrom.ForecastInfo
+	var err error
+	val.AdditionalCost = 0
```
```suggestion
```

andig commented on tariff/ostrom.go:
```diff
@@ -0,0 +1,266 @@
+			}
+			mergeRates(t.data, data)
+		} else {
+			t.log.ERROR.Println(err)
```
Das sieht falsch aus. Fehler zurück geben?

kscholty replied:
Das würde die Run Funktion ja beenden. Mein Gedanke war, dass es diesmal einen Fehler gegeben haben kann, dass aber den nächste Aufruf vielleicht funktioniert.

andig replied:
> Das würde die Run Funktion ja beenden.

Warum? Dafür ist doch der Channel da!

kscholty replied:
Ok, erledigt.

andig commented on tariff/ostrom.go:
```diff
@@ -0,0 +1,266 @@
+
+	tick := time.NewTicker(time.Hour)
+	for ; true; <-tick.C {
+		val.Marketprice, err = t.getFixedPrice()
```
Wozu braucht es val.Marketprice?

kscholty replied:
Der Gesamtpreis ergibt sich als "Marketprice+AdditionalCost", siehe `rate()`. Da steht also der Preis drin. Ich wollte so weit es geht dieselben Datenstrukturen und Funktionen für statische und dynamische Preise verwenden, daher wird hier auch für statische Preise ein `ostrom.ForcastInfo` verwendet, um so `rate()` nutzen zu können.

andig commented on tariff/ostrom.go:
```diff
@@ -0,0 +1,266 @@
+	for ; true; <-tick.C {
+		val.Marketprice, err = t.getFixedPrice()
+		if err == nil {
+			val.StartTimestamp = now.BeginningOfDay()
```
Einfacher analog tariff.go umsetzbar.

andig commented on tariff/ostrom.go:
```diff
@@ -0,0 +1,266 @@
+			data := make(api.Rates, 0, 48)
+			for i := 0; i < 48; i++ {
+				data = append(data, rate(val))
+				val.StartTimestamp = val.StartTimestamp.Add(time.Hour)
```
Wozu braucht es val.StartTimestamp? Das sollte eine lokale Variable sein.

kscholty replied:
Das wird in `rate()` zur  Berechnung von start-und Endzeit verwendet. Ansonsten, siehe mein Kommentar zu "Marketprice"

andig replied:
Wie wärs das umzudrehen und die notwendigen Parameter zu übergeben statt sie in ein Struct zu packen?

kscholty replied:
Habe die Funktion jetzt ganz entfernt, da sie nicht mehr nützlich wäre, wenn ich die Parameter sowieso übergebe. Das Ergebnis wird jetzt dierekt in der Schleife berechnet.

andig commented on tariff/ostrom.go:
```diff
@@ -0,0 +1,266 @@
+		}
+	}
+
+	return 0, errors.New("Could not find basic tariff in tariff response")
```
```suggestion
	return 0, errors.New("tariff not found")
```

andig commented on tariff/ostrom.go:
```diff
@@ -0,0 +1,266 @@
+	if err := t.GetJSON(uri, &city); err != nil {
+		return 0, err
+	}
+	if len(city) == 0 {
+		// Shouldn't happen, but who knows.
+		return 0, fmt.Errorf("No CityId found for zip: %s", t.zip)
+	} else {
+		return city[0].Id, nil
+	}
```
```suggestion
	if len(city) == 0 {
		return 0, errors.New("city not found")
	}

	return city[0].Id, nil
```

kscholty replied:
Ich weiß nicht, ob != 1 nicht etwas zu strikt ist. Zur Vermeidung eines Laufzeitfehlers reicht doch ein Eintrag.

andig commented on tariff/ostrom/api.go:
```diff
@@ -0,0 +1,95 @@
+
+type Tariffs struct {
+	Ostrom []struct {
+		ProductCode                              string  `json:"productCode"`
```
Die json Tags können für lesende Datenstrukturen komplett raus solange das Feld wir das json Tag heisst.

kscholty replied:
Aber schaden die denn? Ich war immer der Meinung, das sei sogar effizienter.
Außerdem ist das automatisch generiert und die wieder alle zu löschen....

kscholty commented on tariff/ostrom.go:
```diff
@@ -0,0 +1,266 @@
+	})
+
+	var res oauth2.Token
+	client := request.NewHelper(t.log)
```
Weil der `t.Client.transport` ein `oauth2.Transport` ist, den ich nicht verwenden kann, um für ihn selber ein Token anzufordern, denn das verklemmt sich dann.
Gibt es eine Alternative, an einen "neutralen" client zu kommen?

andig replied:
Argh, nee…

kscholty committed:
Implemented change requests
Tested in Snadbox with both types of contracts
and in production with static tariff

kscholty committed:
fixed lint error

andig commented on tariff/ostrom.go:
```diff
@@ -0,0 +1,275 @@
+	var cc struct {
+		ClientId     string
+		ClientSecret string
+		Contract     string
```
Int analog api?

kscholty replied:
Es tut mir Leid, aber ich verstehe nicht, was das bedeuten soll.

andig replied:
Im Api ist das ein int. Warum dann hier ein string?

kscholty replied:
Ach so. Ja. Ich fand es schöner, wenn kein Wert angegeben ist. Habe es jetzt auf int64 geändert.

kscholty committed:
Changed ContractId in config to int64

kscholty committed:
Merge branch 'master' into tariff/ostrom

kscholty committed:
Merge branch 'evcc-io:master' into tariff/ostrom

fischer-duelmen commented:
Ich warte auch schon seit September. Werde mich doch wohl mit der Ostromeigenen App beschäftigen müssen. Schade, habe es bisher gerne verwendet.

GrimmiMeloni commented:
Macht euch mal locker, @kscholty und @andig sind doch offensichtlich im Review.

kscholty commented:
Ich hoffe, @andig gibt so sein OK, ich würde das 
so lassen.

kscholty committed:
Merge branch 'master' into tariff/ostrom

andig commented:
Die Review Comments müssten noch bearbeitet werden.

kscholty committed:
Merge branch 'evcc-io:master' into tariff/ostrom

kscholty committed:
Implemented suggested changes

kscholty committed:
Fixed index problem

kscholty commented:
Jetzt sollte alles erledigt sein, @andig. Ich hoffe, so it's ok.

andig commented on tariff/ostrom.go:
```diff
@@ -0,0 +1,276 @@
+	t := &Ostrom{
+		log:          log,
+		basic:        basic,
+		contractType: ostrom.PRODUCT_DYNAMIC,
```
why is this needed?

kscholty replied:
This is needed to implement the api.TarrifType interface.
Ostrom offers two different types of contracts, one with dynamic and one with (more or less) static prices. This stores the information which one it is.

andig replied:
Doesnt that already happen in https://github.com/evcc-io/evcc/pull/16354/files#diff-91620eb465981824c351f0d81e5ab643b5b3751232cce50d8ee91d541db1091eR101?

kscholty replied:
Sorry, but I don't understand which part of the code you refer to.
The Function `Type()` uses the value stored in this member to decide on the type of the tariff. The initialization here is not strictly necessary, since the value will be set after the contracts have been read.

andig committed:
Simplify

andig committed:
Panic on invalid type

andig committed:
Simplify

kscholty committed:
Merge branch 'evcc-io:master' into tariff/ostrom

kscholty commented on tariff/ostrom.go:
```diff
@@ -185,29 +186,28 @@ func (t *Ostrom) GetContracts() ([]ostrom.Contract, error) {
+
+		data := make(api.Rates, 48)
+		for i := range data {
+			ts := now.BeginningOfDay().Add(time.Duration(i) * time.Hour)
```
Ich verstehe ja, dass unsere Auffassungen von gutem Code sich unterscheiden.
Aber warum ist es besser, hier 96 Funktionsaufrufe und 48 Multiplikationen zu machen, nur um Werte zu berechnen, die du in der jeweils vorhergehenden Iteration bereits berechnet hast?
Es gilt doch data[i].Start = data[i-1].End

andig replied:
Hilft uns denn die Diskussion? Am Ende muss das Coreteam den Code warten und Geschmack ist subjektiv.

andig replied:
Im Prinzip ja, aber so eine Multiplikation (oder 48) alle 2 Stunden interessiert schon lange keine CPU mehr. Dann lieber lesbar, Stichwort „premature optimization“. Das ist aber sicher kein Hügel, auf dem ich sterben möchte 😅

andig commented on tariff/ostrom.go:
```diff
@@ -0,0 +1,266 @@
+			continue
+		}
+
+		data := make(api.Rates, 48)
```
Hier vllt. noch mit append arbeiten falls die Anzahl nicht stimmt! Stichwort Zeitumstellung!!!

kscholty replied:
Ok, guter Hinweis

andig commented on tariff/ostrom.go:
```diff
@@ -0,0 +1,266 @@
+	t := &Ostrom{
+		log:          log,
+		basic:        basic,
+		contractType: ostrom.PRODUCT_DYNAMIC,
```
```suggestion
```

andig replied:
Dann raus damit wenns unnötig ist 👍🏻

kscholty committed:
Using append for dynamic tarriffs.

kscholty committed:
Update tariff/ostrom.go

Co-authored-by: andig <cpuidle@gmail.com>

kscholty committed:
Removed lint problem

andig commented:
Danke für Deine Geduld, sieht super aus!

tipuraneo commented:
In der [EVCC-Doku](https://docs.evcc.io/docs/devices/tariffs#ostrom) steht:
"Erzeuge einen 'Production Client' in dem Tibber-Entwicklerportal:"
Hier müsste "Tibber" durch "Ostrom" ersetzt werden.
